### PR TITLE
ENH: Simple COO format

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @hameerabbasi,

This PR features simple COO (2dim) format. Some comments:
- It's still with a lot of boilerplate that is repeated in dense and csr formats, I still need to figure out proper abstraction here.
- I think there's something wrong with 2dim case of `MemrefType` strides. I did `index = index.flatten().reshape((pos[1], 2), order="F").T` hack which should go away, as it looks like for `(25, 2)` shape the strides are `(2, 1)`. Or the strides are correct and `aligned` content is incorrect (looks like is columnwise for some reason).
- I'm afraid our `sparse.add` implementation isn't generic, as `coo+csr` that I added in tests fails.

### CODE REVIEW

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
It seems like you're right about the strides, they should be some variant of `25 * 8` in there somewhere.

**Comment by hameerabbasi** [comment]
Also -- MLIR strides are measured in items; NumPy strides are measured in bytes.

File: sparse/mlir_backend/_constructors.py [hunk 1]
```diff
@@ -141,8 +141,107 @@ def to_np(self) -> np.ndarray:
...
+        module = mlir.execution_engine.ExecutionEngine(module, opt_level=2, shared_libs=[MLIR_C_RUNNER_UTILS])
+        return (module, coo_shaped)
+
+    @classmethod
+    def assemble(cls, module: ir.Module, arr: sps.coo_array) -> ctypes.c_void_p:
+        out = ctypes.c_void_p()
+        module.invoke(
+            "assemble",
+            ctypes.pointer(ctypes.pointer(ranked_memref_from_np(np.array([0, arr.size])))),
+            ctypes.pointer(ctypes.pointer(ranked_memref_from_np(np.copy(np.vstack(arr.coords).T)))),
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply sparse/mlir_backend/_constructors.py]: Done!

File: sparse/mlir_backend/_constructors.py [hunk 2]
```diff
@@ -141,8 +141,107 @@ def to_np(self) -> np.ndarray:
...
+                @func.FuncOp.from_py_func(tensor_1d_index, tensor_2d_index, tensor_1d_values)
+                def assemble(pos, index, values):
+                    return sparse_tensor.assemble(coo_shaped, (pos, index), values)
+
+                @func.FuncOp.from_py_func(coo_shaped)
+                def disassemble(tensor_shaped):
+                    nse = sparse_tensor.number_of_entries(tensor_shaped)
+                    pos = tensor.EmptyOp([arith.constant(ir.IndexType.get(), 0)], index_dtype)
+                    index = tensor.EmptyOp([nse, 2], index_dtype)
+                    values = tensor.EmptyOp([arith.constant(ir.IndexType.get(), 0)], values_dtype)
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply sparse/mlir_backend/_constructors.py]: Done!

File: sparse/mlir_backend/_constructors.py [hunk 3]
```diff
@@ -141,8 +141,107 @@ def to_np(self) -> np.ndarray:
...
+            with ir.InsertionPoint(module.body):
+
+                @func.FuncOp.from_py_func(tensor_1d_index, tensor_2d_index, tensor_1d_values)
+                def assemble(pos, index, values):
+                    return sparse_tensor.assemble(coo_shaped, (pos, index), values)
+
+                @func.FuncOp.from_py_func(coo_shaped)
+                def disassemble(tensor_shaped):
+                    nse = sparse_tensor.number_of_entries(tensor_shaped)
+                    pos = tensor.EmptyOp([arith.constant(ir.IndexType.get(), 0)], index_dtype)
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply sparse/mlir_backend/_constructors.py]: Done!

**Comment by mtsokol** [comment]
> It seems like you're right about the strides, they should be some variant of `25 * 8` in there somewhere.

I think the issue was with our custom `_memref.py` implementation. I couldn't figure out where the issue was, but it turns out there's `import mlir.runtime as rt` module that already has functions for moving NumPy arrays from and to MLIR.

File: sparse/mlir_backend/tests/test_simple.py [hunk 4]
```diff
@@ -115,3 +120,10 @@ def test_add(rng, dtype):
     expected = a + c
     assert isinstance(actual, np.ndarray)
     np.testing.assert_array_equal(actual, expected)
+
+    # TODO: Blocked by https://github.com/llvm/llvm-project/issues/107477
+    # d = sps.random_array(SHAPE, density=DENSITY, format="coo", dtype=dtype, random_state=rng)
+    # d_tensor = sparse.asarray(d)
+    # actual = sparse.add(b_tensor, d_tensor).to_scipy_sparse()
+    # expected = b + d
+    # np.testing.assert_array_equal(actual.todense(), expected.todense())
```
Comment (hameerabbasi) [inline sparse/mlir_backend/tests/test_simple.py:129]: Does this still not work?

Comment (mtsokol) [reply sparse/mlir_backend/tests/test_simple.py:129]: Yes, adding COO tensors and returning the result as COO is broken IMO and explained in detail here: https://github.com/llvm/llvm-project/issues/107477

**Review by hameerabbasi — APPROVED** [review]
LGTM!

**Pull request merged**